external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutEmitTest.java

```diff
@@ -134,110 +130,86 @@ public void testNextTupleEmitsFailedMessagesEvenWhenMaxUncommittedOffsetsIsExcee
+            setupSpout(partitions);
+            Map<TopicPartition, List<ConsumerRecord<String, String>>> records = new HashMap<>();
+            //This is cheating a bit since maxPollRecords would normally spread this across multiple polls
+            records.put(partition, createRecords(partition, 0, spoutConfig.getMaxUncommittedOffsets()));
+            records.put(partitionTwo, createRecords(partitionTwo, 0, spoutConfig.getMaxUncommittedOffsets()));
 
             when(consumerMock.poll(anyLong()))
-                .thenReturn(new ConsumerRecords(firstPollRecords))
-                .thenReturn(new ConsumerRecords(secondPollRecords));
+                .thenReturn(new ConsumerRecords(records));
 
-            for (int i = 0; i < spoutConfig.getMaxUncommittedOffsets() + maxPollRecords; i++) {
+            for (int i = 0; i < spoutConfig.getMaxUncommittedOffsets()*2; i++) {
                 spout.nextTuple();
             }
 
             ArgumentCaptor<KafkaSpoutMessageId> messageIds = ArgumentCaptor.forClass(KafkaSpoutMessageId.class);
-            verify(collectorMock, times(firstPollRecordsForPartition.size())).emit(anyObject(), anyObject(), messageIds.capture());
-
-            KafkaSpoutMessageId failedMessageId = messageIds.getAllValues().get(messageIds.getAllValues().size() - 1);
-            spout.fail(failedMessageId);
-
+            verify(collectorMock, times(spoutConfig.getMaxUncommittedOffsets()*2)).emit(anyObject(), anyObject(), messageIds.capture());
+            
+            //Now fail a tuple on partition 0 and verify that it is allowed to retry
+            //Partition 1 should be paused, since it is at the uncommitted offsets limit
+            Optional<KafkaSpoutMessageId> failedMessageId = messageIds.getAllValues().stream()
+                .filter(messageId -> messageId.partition() == partition.partition())
+                .findAny();
+            
+            spout.fail(failedMessageId.get());
+            
             reset(collectorMock);
-
-            //Now make the single failed tuple retriable
+            
             Time.advanceTime(50);
-            //The spout should allow another poll since there are now only maxUncommittedOffsets - 1 nonretriable tuples
-            for (int i = 0; i < firstPollRecordsForPartition.size() + maxPollRecords; i++) {
-                spout.nextTuple();
-            }
-
-            ArgumentCaptor<KafkaSpoutMessageId> retryBatchMessageIdsCaptor = ArgumentCaptor.forClass(KafkaSpoutMessageId.class);
-            verify(collectorMock, times(maxPollRecords)).emit(anyObject(), anyObject(), retryBatchMessageIdsCaptor.capture());
-            reset(collectorMock);
             
-            //Check that the consumer started polling at the failed tuple offset
+            when(consumerMock.poll(anyLong()))
+                .thenReturn(new ConsumerRecords(Collections.singletonMap(partition, createRecords(partition, failedMessageId.get().offset(), 1))));
```

Comment:
return generic ConsumerRecords to avoid warning around raw type

In this case I am getting an incompatible bounds error on the V param to singletonMap()